internal/handlers/user.go

```diff
@@ -39,63 +42,93 @@ func (api *API) UserIsCurrentUser(c *gin.Context) func(db *gorm.DB) *gorm.DB {
 	}
 }
 
-var noUUID = uuid.UUID{}
-
 func (api *API) createUserIfNotExists(ctx context.Context, id string, userName string) (uuid.UUID, error) {
 	ctx, span := tracer.Start(ctx, "createUserIfNotExists")
 	defer span.End()
 	span.SetAttributes(
 		attribute.String("user-id", id),
 		attribute.String("username", userName),
 	)
-	tx := api.db
 	var user models.User
-	res := tx.Unscoped().First(&user, "id = ?", id)
-	if res.Error == nil {
-
-		// Was the user was previously deleted... lets make him active again.
-		if user.DeletedAt.Valid {
-			user.DeletedAt = gorm.DeletedAt{}
-			res = tx.Unscoped().Model(&user).Update("DeletedAt", user.DeletedAt)
-			if res.Error != nil {
-				return noUUID, res.Error
+	var uuid uuid.UUID
+
+	err := api.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
+		// First lets check if the user has ever existed in the database
+		res := tx.Unscoped().First(&user, "id = ?", id)
+
+		// If the user exists, then lets restore their status in the database
+		if res.Error == nil {
+			var err error
+			uuid, err = api.restoreDeletedUser(ctx, tx, &user, id, userName)
+			if err != nil {
+				return err
 			}
+
+			return nil
 		}
 
-		// Check if the UserName has changed since the last time we saw this user
-		if user.UserName != userName {
-			res = tx.Model(&user).Update("UserName", userName)
-			if res.Error != nil {
-				return noUUID, res.Error
-			}
+		if !errors.Is(res.Error, gorm.ErrRecordNotFound) {
+			return fmt.Errorf("can't find record for user id %s: %w", id, res.Error)
 		}
 
-		return api.createUserOrgIfNotExists(ctx, id, userName)
+		user.ID = id
+		user.UserName = userName
+		if res = tx.Create(&user); res.Error != nil {
+			return res.Error
+		}
+		var err error
+		uuid, err = api.createUserOrgIfNotExists(ctx, tx, id, userName)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return noUUID, fmt.Errorf("can't create user record: %w", err)
 	}
 
-	if !errors.Is(res.Error, gorm.ErrRecordNotFound) {
-		return noUUID, fmt.Errorf("can't find record for user id %s", id)
+	if uuid != noUUID {
+		return uuid, nil
 	}
-	user.ID = id
-	user.UserName = userName
-	res = tx.Create(&user)
-	if res.Error == nil {
-		return api.createUserOrgIfNotExists(ctx, id, userName)
+
+	return noUUID, fmt.Errorf("can't create user record")
+}
+
+// restoreDeleteUser will restore a user if they have been deleted
```

Comment:
Will fix this up in next commit that will incorporate feedback on other commit as well as general PR feedback